### PR TITLE
Make count with skip to work with CouchDB

### LINKED ISSUE #3560: Count with skip is not working for CouchDB
As mentioned [here][1] in #3517 count with skip seems to be not working for CouchDB. As a fix we can instead compute the count by subtracting skip from overall count in CouchDBReststore. 

Once this is fixed the test case `ArtifactStoreQueryBehaviors` should be enabled again

[1]: https://github.com/apache/incubator-openwhisk/pull/3517#issuecomment-381031566

### PR DESCRIPTION
**Pull request by chetanmeh**
Fixes #3560

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3560)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [x] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by chetanmeh** [comment]
@rabbah Please review

**Comment by rabbah** [comment]
@csantanapr @dubee do you mind a PG 🚀 for this? It should be OK but better to be sure.

**Comment by csantanapr** [comment]
PG2 3080 🏃

**Comment by csantanapr** [comment]
PG2 3081 🔵

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala [hunk 1]
```diff
@@ -315,19 +315,14 @@ class CouchDbRestStore[DocumentAbstraction <: DocumentSerializer](dbProtocol: St
...
-        reduce = true)
+      .executeView(firstPart, secondPart)(startKey = startKey, endKey = endKey, stale = stale, reduce = true)
       .map {
         case Right(response) =>
           val rows = response.fields("rows").convertTo[List[JsObject]]
 
           val out = if (!rows.isEmpty) {
             assert(rows.length == 1, s"result of reduced view contains more than one value: '$rows'")
-            rows.head.fields("value").convertTo[Long]
+            rows.head.fields("value").convertTo[Long] - skip
```
Comment (dubee) [inline common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: @chetanmeh, what happens if skip > value?

Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Good point!. That case was not handled and would have resulted in negative count. Updated that to return 0 and also added a test to check that scenario

Comment (rabbah) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: nice catch... shame on me for missing it.

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala [hunk 2]
```diff
@@ -315,19 +315,15 @@ class CouchDbRestStore[DocumentAbstraction <: DocumentSerializer](dbProtocol: St
...
       .map {
         case Right(response) =>
           val rows = response.fields("rows").convertTo[List[JsObject]]
 
-          val out = if (!rows.isEmpty) {
+          val out = if (rows.nonEmpty) {
             assert(rows.length == 1, s"result of reduced view contains more than one value: '$rows'")
-            rows.head.fields("value").convertTo[Long]
+            val count = rows.head.fields("value").convertTo[Long]
+            if (count > skip) count - skip else 0L
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: skip may be negative also - although it is better to reject.a negative skip at the API handler (validate the value there) but defensively code against that as well here?

Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Makes sense. Added a check for both `limit` and `skip` to  be non negative in both `query` and `count`

**Comment by rabbah** [comment]
@dubee https://github.com/apache/incubator-openwhisk-cli/issues/187

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala (hunk 2 above)
Comment (dubee) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: @rabbah, are you saying to terminate a route if those values are negative at the top API level?

Comment (rabbah) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: yes - we can just validate and reject early (we do some other validations, so could extend to count and skip)... orthogonal to this PR.

Comment (dubee) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Make sense. With these changes, `There was an internal server error.` is returned when a negative `skip` value is specified. A content-type of `text/plain; charset=UTF-8` is returned as well.

**Pull request merged**